### PR TITLE
improve Mat::operator=(Scalar)

### PR DESCRIPTION
**Pull request by nhlsm**
improve Mat::operator=(Scalar)
- Mat::zeros() is faster than Mat::ones(). 
- The reason is that Mat::zeros() use memset(), but Mat::ones() use memcpy() for every row.
- As a result, Mat::ones() need more memory access than Mat::zeros().
-  I modified Mat::operator=(Scalar) that use memset() on possible case. ( 8U, 8S, and 8UC3 same scalar element etc...)
- 50% performance enhancement

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under OpenCV (BSD) License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [X] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
Thank you for contribution!

This patch should go into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
We will merge changes from 3.4 into master regularly (weekly/bi-weekly).

So, please:
- change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
- rebase your commits from master onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/master`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no needs to re-open PR, apply changes "inplace".

File: modules/core/src/copy.cpp [hunk 1]
```diff
@@ -414,6 +414,46 @@ void Mat::copyTo( OutputArray _dst, InputArray _mask ) const
         copymask(ptrs[0], 0, ptrs[2], 0, ptrs[1], 0, sz, &esz);
 }
 
+// just 1 byte is valid
+#define CANT_APPLY_MEMSET int(0xDEADBEEF)
+
+static int check_apply_memset(const Mat *mat, const Scalar &s)
```
Comment (alalek) [inline modules/core/src/copy.cpp:420]:
```.diff
-const Mat *mat,
+const Mat& mat,
```

to make static analyzers happy (missing unnecessary zero pointer check)

File: modules/core/src/copy.cpp [hunk 2]
```diff
@@ -414,6 +414,46 @@ void Mat::copyTo( OutputArray _dst, InputArray _mask ) const
...
+{
+    int fill_value = CANT_APPLY_MEMSET;
+    switch (mat->depth())
+    {
+    case CV_8U: fill_value = saturate_cast<uchar>( s.val[0] ); break;
+    case CV_8S: fill_value = saturate_cast<schar>( s.val[0] ); break;
+    default:
+        //CV_Error(CV_BadDepth, "Unsupported depth");
```
Comment (alalek) [inline modules/core/src/copy.cpp:428]: Remove dead code

File: modules/core/src/copy.cpp [hunk 3]
```diff
@@ -414,6 +414,46 @@ void Mat::copyTo( OutputArray _dst, InputArray _mask ) const
         copymask(ptrs[0], 0, ptrs[2], 0, ptrs[1], 0, sz, &esz);
 }
 
+// just 1 byte is valid
+#define CANT_APPLY_MEMSET int(0xDEADBEEF)
```
Comment (alalek) [inline modules/core/src/copy.cpp:418]: Avoid magic constants.
- return "bool" value
- add "output" parameter reference which will be filled.

(keep source code and binaries clean)

File: modules/core/src/copy.cpp (hunk 1 above)
Comment (nhlsm) [reply modules/core/src/copy.cpp:420]: I agree.

File: modules/core/src/copy.cpp (hunk 3 above)
Comment (nhlsm) [reply modules/core/src/copy.cpp:418]: I agree. ( nice suggestion )
Your suggestion make code better clean.

File: modules/core/src/copy.cpp (hunk 2 above)
Comment (nhlsm) [reply modules/core/src/copy.cpp:428]: I do not agree.
- for readability : code reader save time without reading unnecessary code below.
- for short-circuit : remove unnecessary operation.

**Pull request merged**